File: tyhp-lang/phpstorm/src/main/kotlin/com/tyhp/lang/debug/ProxyProcessController.kt
package com.tyhp.lang.debug

import java.util.concurrent.CopyOnWriteArrayList
import java.util.concurrent.CompletableFuture
import java.util.concurrent.ExecutorService
import java.util.concurrent.Executors
import java.util.concurrent.TimeUnit
import java.util.concurrent.atomic.AtomicBoolean

/**
 * Minimal child-process surface used by the proxy controller. Tests inject
 * fakes; the manager wraps `Process`.
 */
interface SpawnedProxy {
    val pid: Long?
    val exitCode: Int?
    fun onExit(listener: (Int?) -> Unit)
    fun onError(listener: (Throwable) -> Unit)
    /** Manager stop path — JVM `Process.destroy()` (SIGTERM on Unix), not a shell `kill`. */
    fun destroyTerm(): Boolean
}

data class ProxyStartRequest(
    val command: String,
    val args: List<String>,
    val cwd: String? = null,
    val idePort: Int,
)

interface ProxyProcessHooks {
    fun spawn(command: String, args: List<String>, cwd: String?): SpawnedProxy
    fun waitForListening(port: Int, child: SpawnedProxy, abort: AtomicBoolean): Int
    fun stopProcess(child: SpawnedProxy)
    fun onLog(line: String) {}
    fun onState(state: ProxyRunState) {}
    fun onUnexpectedExit(code: Int?) {}
}

const val PROXY_STOP_TIMEOUT_MS = 8_000L
const val PROXY_LISTEN_TIMEOUT_MS = 15_000L

/**
 * Serializes start/stop/restart and applies [nextProxyState]. IntelliJ-free so
 * the state machine can be unit-tested as a plain JVM test.
 */
class ProxyProcessController(
    private val hooks: ProxyProcessHooks,
    private val executor: ExecutorService = Executors.newSingleThreadExecutor { runnable ->
        Thread(runnable, "tyhp-xdebug-proxy").apply { isDaemon = true }
    },
) {
    private val lock = Any()

    @Volatile
    var currentState: ProxyRunState = ProxyRunState.STOPPED
        private set

    @Volatile
    private var child: SpawnedProxy? = null

    @Volatile
    private var expectedStop = false

    private val startAbort = AtomicBoolean(false)

    val currentPid: Long?
        get() = child?.pid

    fun start(request: ProxyStartRequest): CompletableFuture<Void> =
        enqueue { startImpl(request) }

    fun stop(): CompletableFuture<Void> =
        enqueue { stopImpl() }

    fun restart(request: ProxyStartRequest): CompletableFuture<Void> =
        enqueue {
            stopImpl()
            startImpl(request)
        }

    fun shutdownExecutor() {
        executor.shutdown()
    }

    private fun enqueue(work: () -> Unit): CompletableFuture<Void> {
        return CompletableFuture.runAsync(work, executor)
    }

    private fun startImpl(request: ProxyStartRequest) {
        if (currentState == ProxyRunState.RUNNING || currentState == ProxyRunState.STARTING) {
            return
        }
        setState(nextProxyState(currentState, ProxyLifecycleEvent.StartRequested))
        expectedStop = false
        startAbort.set(false)
        try {
            val spawned = hooks.spawn(request.command, request.args, request.cwd)
            child = spawned
            spawned.onExit { code -> handleExit(code) }
            spawned.onError { err ->
                hooks.onLog("Proxy process error: ${err.message ?: err}")
            }
            val boundPort = hooks.waitForListening(request.idePort, spawned, startAbort)
            if (startAbort.get()) {
                stopImpl()
                return
            }
            hooks.onLog("Proxy listening on IDE port $boundPort.")
            setState(nextProxyState(currentState, ProxyLifecycleEvent.Listening))
        } catch (err: Throwable) {
            val message = err.message ?: err.toString()
            hooks.onLog("Proxy failed to start: $message")
            setState(nextProxyState(currentState, ProxyLifecycleEvent.StartFailed))
            cleanupChild(keepError = true)
            throw err
        }
    }

    private fun stopImpl() {
        cleanupChild(keepError = false)
    }

    private fun cleanupChild(keepError: Boolean) {
        startAbort.set(true)
        val spawned = child
        if (spawned == null) {
            if (!keepError) {
                setState(ProxyRunState.STOPPED)
            }
            return
        }
        expectedStop = true
        if (!keepError) {
            setState(nextProxyState(currentState, ProxyLifecycleEvent.StopRequested))
        }
        try {
            withTimeout(PROXY_STOP_TIMEOUT_MS, "Proxy did not exit after SIGTERM") {
                hooks.stopProcess(spawned)
            }
        } catch (err: Throwable) {
            hooks.onLog(err.message ?: err.toString())
            setState(ProxyRunState.ERROR)
            child = null
            throw err
        }
        child = null
        if (!keepError) {
            setState(nextProxyState(currentState, ProxyLifecycleEvent.Exited(expected = true)))
        }
    }

    private fun handleExit(code: Int?) {
        val expected = expectedStop
        child = null
        setState(nextProxyState(currentState, ProxyLifecycleEvent.Exited(expected)))
        if (!expected && currentState == ProxyRunState.ERROR) {
            hooks.onUnexpectedExit(code)
        }
    }

    private fun setState(state: ProxyRunState) {
        synchronized(lock) {
            if (currentState == state) {
                return
            }
            currentState = state
        }
        hooks.onState(state)
    }
}

fun waitForChildExit(child: SpawnedProxy, timeoutMs: Long) {
    if (child.exitCode != null) {
        return
    }
    val done = CompletableFuture<Void>()
    child.onExit { done.complete(null) }
    if (child.exitCode != null) {
        return
    }
    try {
        done.get(timeoutMs, TimeUnit.MILLISECONDS)
    } catch (err: Exception) {
        throw IllegalStateException("Proxy process did not exit within ${timeoutMs}ms", err)
    }
}

/**
 * Manager stop path: signal the child with SIGTERM (`Process.destroy`) and wait
 * for exit. Does not use shell `kill` / `killall`.
 */
fun stopSpawnedProxy(child: SpawnedProxy) {
    if (child.exitCode != null) {
        return
    }
    val done = CompletableFuture<Void>()
    child.onExit { done.complete(null) }
    if (child.exitCode != null) {
        return
    }
    child.destroyTerm()
    try {
        done.get(PROXY_STOP_TIMEOUT_MS, TimeUnit.MILLISECONDS)
    } catch (err: Exception) {
        throw IllegalStateException("Proxy process did not exit within ${PROXY_STOP_TIMEOUT_MS}ms", err)
    }
}

private fun withTimeout(ms: Long, message: String, work: () -> Unit) {
    val future = CompletableFuture.runAsync(work)
    try {
        future.get(ms, TimeUnit.MILLISECONDS)
    } catch (err: Exception) {
        throw IllegalStateException(message, err)
    }
}

/** Test helper: an in-memory child that records `destroyTerm` and can be exited. */
class FakeSpawnedProxy : SpawnedProxy {
    override var pid: Long? = 4242
    override var exitCode: Int? = null
        private set

    var killed: Boolean = false
        private set
    var lastSignal: String? = null
        private set

    private val exitListeners = CopyOnWriteArrayList<(Int?) -> Unit>()
    private val errorListeners = CopyOnWriteArrayList<(Throwable) -> Unit>()

    override fun onExit(listener: (Int?) -> Unit) {
        exitListeners.add(listener)
        val code = exitCode
        if (code != null) {
            listener(code)
        }
    }

    override fun onError(listener: (Throwable) -> Unit) {
        errorListeners.add(listener)
    }

    override fun destroyTerm(): Boolean {
        killed = true
        lastSignal = "SIGTERM"
        exit(0)
        return true
    }

    fun exit(code: Int? = 0) {
        if (exitCode != null) {
            return
        }
        exitCode = code
        for (listener in exitListeners) {
            listener(code)
        }
    }
}

File: tyhp-lang/vscode/src/debug/ProxyProcessController.ts
import { EventEmitter } from "node:events";
import { nextProxyState, ProxyRunState } from "./proxyLifecycle";

/**
 * Minimal child-process surface used by the proxy controller. Tests inject
 * fakes; the manager wraps `child_process.spawn`.
 */
export interface SpawnedProxy {
    readonly pid?: number;
    readonly stdout?: NodeJS.ReadableStream;
    readonly stderr?: NodeJS.ReadableStream;
    readonly exitCode: number | null;
    on(event: "exit" | "error", listener: (...args: unknown[]) => void): void;
    /** Manager stop path — Node `ChildProcess.kill`, not a shell `kill`. */
    kill(signal?: NodeJS.Signals): boolean;
}

export interface ProxyStartRequest {
    command: string;
    args: readonly string[];
    cwd?: string;
    idePort: number;
}

export interface ProxyProcessHooks {
    spawn(command: string, args: readonly string[], cwd?: string): SpawnedProxy;
    waitForListening(
        port: number,
        child: SpawnedProxy,
        abort: AbortSignal
    ): Promise<number>;
    stopProcess(child: SpawnedProxy): Promise<void>;
    onLog?(line: string): void;
    onState?(state: ProxyRunState): void;
    onUnexpectedExit?(code: number | null): void;
    onNoSourcemaps?(): void;
}

const STOP_TIMEOUT_MS = 8_000;

/**
 * Serializes start/stop/restart and applies {@link nextProxyState}. Does not
 * import `vscode` so the state machine can be unit-tested under `node --test`.
 */
export class ProxyProcessController {
    private state: ProxyRunState = "stopped";
    private child: SpawnedProxy | undefined;
    private expectedStop = false;
    private startAbort: AbortController | undefined;
    private queue: Promise<void> = Promise.resolve();
    private readonly exitBound = (code: unknown, _signal: unknown) => {
        this.handleExit(typeof code === "number" ? code : null);
    };

    constructor(private readonly hooks: ProxyProcessHooks) {}

    get currentState(): ProxyRunState {
        return this.state;
    }

    get currentPid(): number | undefined {
        return this.child?.pid;
    }

    start(request: ProxyStartRequest): Promise<void> {
        return this.enqueue(() => this.startImpl(request));
    }

    stop(): Promise<void> {
        return this.enqueue(() => this.stopImpl());
    }

    restart(request: ProxyStartRequest): Promise<void> {
        return this.enqueue(async () => {
            await this.stopImpl();
            await this.startImpl(request);
        });
    }

    private enqueue(work: () => Promise<void>): Promise<void> {
        const run = this.queue.then(work, work);
        this.queue = run.then(
            () => undefined,
            () => undefined
        );
        return run;
    }

    private async startImpl(request: ProxyStartRequest): Promise<void> {
        if (this.state === "running" || this.state === "starting") {
            return;
        }
        this.setState(nextProxyState(this.state, { type: "startRequested" }));
        this.expectedStop = false;
        this.startAbort = new AbortController();
        try {
            const child = this.hooks.spawn(request.command, request.args, request.cwd);
            this.child = child;
            child.on("exit", this.exitBound);
            child.on("error", (err: unknown) => {
                const message = err instanceof Error ? err.message : String(err);
                this.hooks.onLog?.(`Proxy process error: ${message}`);
            });
            const boundPort = await this.hooks.waitForListening(
                request.idePort,
                child,
                this.startAbort.signal
            );
            if (this.startAbort.signal.aborted) {
                await this.stopImpl();
                return;
            }
            this.hooks.onLog?.(`Proxy listening on IDE port ${boundPort}.`);
            this.setState(nextProxyState(this.state, { type: "listening" }));
        } catch (err) {
            const message = err instanceof Error ? err.message : String(err);
            this.hooks.onLog?.(`Proxy failed to start: ${message}`);
            this.setState(nextProxyState(this.state, { type: "startFailed" }));
            await this.cleanupChild({ keepError: true });
            throw err;
        } finally {
            this.startAbort = undefined;
        }
    }

    private async stopImpl(): Promise<void> {
        await this.cleanupChild({ keepError: false });
    }

    private async cleanupChild(options: { keepError: boolean }): Promise<void> {
        this.startAbort?.abort();
        this.startAbort = undefined;
        const child = this.child;
        if (!child) {
            if (!options.keepError) {
                this.setState("stopped");
            }
            return;
        }
        this.expectedStop = true;
        if (!options.keepError) {
            this.setState(nextProxyState(this.state, { type: "stopRequested" }));
        }
        try {
            await withTimeout(
                this.hooks.stopProcess(child),
                STOP_TIMEOUT_MS,
                "Proxy did not exit after SIGTERM"
            );
        } catch (err) {
            const message = err instanceof Error ? err.message : String(err);
            this.hooks.onLog?.(message);
            this.setState("error");
            this.child = undefined;
            throw err;
        }
        this.child = undefined;
        if (!options.keepError) {
            this.setState(nextProxyState(this.state, { type: "exited", expected: true }));
        }
    }

    private handleExit(code: number | null): void {
        const expected = this.expectedStop;
        this.child = undefined;
        this.setState(nextProxyState(this.state, { type: "exited", expected }));
        if (!expected && this.state === "error") {
            this.hooks.onUnexpectedExit?.(code);
        }
    }

    private setState(state: ProxyRunState): void {
        if (this.state === state) {
            return;
        }
        this.state = state;
        this.hooks.onState?.(state);
    }
}

export function waitForChildExit(child: SpawnedProxy, timeoutMs: number): Promise<void> {
    if (child.exitCode !== null) {
        return Promise.resolve();
    }
    return new Promise((resolve, reject) => {
        const timer = setTimeout(() => {
            reject(new Error(`Proxy process did not exit within ${timeoutMs}ms`));
        }, timeoutMs);
        child.on("exit", () => {
            clearTimeout(timer);
            resolve();
        });
    });
}

/**
 * Manager stop path: signal the child with SIGTERM and wait for `exit`.
 * Does not use shell `kill` / `killall`.
 */
export async function stopSpawnedProxy(child: SpawnedProxy): Promise<void> {
    if (child.exitCode !== null) {
        return;
    }
    child.kill("SIGTERM");
    await waitForChildExit(child, STOP_TIMEOUT_MS);
}

function withTimeout<T>(promise: Promise<T>, ms: number, message: string): Promise<T> {
    return new Promise<T>((resolve, reject) => {
        const timer = setTimeout(() => reject(new Error(message)), ms);
        promise.then(
            (value) => {
                clearTimeout(timer);
                resolve(value);
            },
            (err) => {
                clearTimeout(timer);
                reject(err);
            }
        );
    });
}

/** Test helper: an in-memory child that records `kill` and can be exited. */
export class FakeSpawnedProxy extends EventEmitter implements SpawnedProxy {
    pid = 4242;
    exitCode: number | null = null;
    killed = false;
    lastSignal: NodeJS.Signals | undefined;

    kill(signal?: NodeJS.Signals): boolean {
        this.killed = true;
        this.lastSignal = signal;
        this.exit(0, signal);
        return true;
    }

    exit(code: number | null = 0, signal?: NodeJS.Signals): void {
        if (this.exitCode !== null) {
            return;
        }
        this.exitCode = code;
        this.emit("exit", code, signal);
    }
}
